platform/ui/src/hooks/useMedia.js

```diff
@@ -1,54 +1,223 @@
+ *
+ *  const currentDisplaySize = useDisplayMediaSize();
+ *
+ */
+const useDisplayMediaSize = (
+  mediaQueriesStringList,
+  mediaTypesAliases,
+  defaultMediaType
+) => {
+  // MediaQuery.state is the source of truth. This hook will be dependent on it.
+  const [state, setState] = useState(MediaQuery.state);
+  let mount = useRef(false);
+
+  // bind current hook set state to MediaQuery setState
+  if (!MediaQuery.setters.includes(setState)) {
+    MediaQuery.setters.push(nextState => {
+      // last chance to avoid setState of unmount component
+      if (mount.current) {
+        setState(nextState);
+      }
+    });
+  }
+
+  // update state of MediaQuery in case mediaQueriesStringList or mediaTypesAliases has changed
+  useEffect(() => {
+    const {
+      mediaQueriesStringList: _mediaQueriesStringList,
+      mediaTypesAliases: _mediaTypesAliases,
+    } = state;
+    if (
+      (mediaQueriesStringList &&
+        !isEqual(mediaQueriesStringList, _mediaQueriesStringList)) ||
+      (mediaTypesAliases && !isEqual(mediaTypesAliases, _mediaTypesAliases))
+    ) {
+      MediaQuery.setState({
+        mediaQueriesStringList,
+        mediaTypesAliases,
+        defaultMediaType,
+      });
+    }
+  }, [mediaQueriesStringList, mediaTypesAliases]);
+
+  // re-assign window resizing listeners
+  useEffect(() => {
+    const { mediaQueryMap } = state;
+    addMediaQueryListeners(mediaQueryMap);
+  }, [state.mediaQueryMap]);
+
+  useEffect(() => {
+    mount.current = true;
+
+    return () => {
+      const { mediaQueryMap } = state;
+      removeMediaQueryListeners(mediaQueryMap);
+      MediaQuery.setters = [];
+      mount.current = false;
+    };
+  }, []);
+
+  return state.displaySize;
+};
+/**
+ * Hook to get a content based on current displayMedia size.
+ *
+ * It uses useDisplayMediaSize. Hook immutable on content changing.
+ *
+ * Current hook only offers content, it wont expose method to change its state.
+ *
+ * @param {Object} contentArrayMap - Mapping object for mediaTypesAliases to content (of any type).
+ * @param {Any} defaultContent - Default content to be used in case current displayMedia size not present on contentArrayMap
+ * @returns {Any} current content based on displayMedia size.
+ *
+ * @example <caption>Example to getContent based on displayMedia size</caption>
+ *
+ *    const currentComponent = useDisplayMediaContent({"small": ComponentA, "medium": ComponentB}, ComponentA);
+ *    const currentObject = useDisplayMediaContent({"small": ObjectA, "medium": ObjectB}, ObjectB);
+ */
+const useDisplayMediaContent = (contentArrayMap, defaultContent) => {
+  const displaySize = useDisplayMediaSize();
+
+  const getContent = () => {
+    const content =
+      displaySize in contentArrayMap
+        ? contentArrayMap[displaySize]
+        : defaultContent;
+    return content;
+  };
 
-  return value;
-}
+  return getContent();
+};
 
-export default useMedia;
+export { useDisplayMediaSize, useDisplayMediaContent };
```

Comment:
This feels significantly more complex than the original implementation. Is all of this necessary? Can you walk me through:

- Issues with the original `useMedia` implementation
- Issues with the second iteration
- How the current solution is optimal for solving `useMedia`'s original intent?